tests/test_expectations.py

```diff
@@ -0,0 +1,189 @@
+            return quad_impl(params[0], params[1], None, None, None)
+    elif len(params) == 3:
+        if isinstance(params[1], tuple) and isinstance(params[2], tuple):
+            return quad_impl(params[0], *params[1], *params[2])
+        elif isinstance(params[1], tuple) and not isinstance(params[2], tuple):
+            return quad_impl(params[0], *params[1], params[2], None)
+        elif not isinstance(params[1], tuple) and isinstance(params[2], tuple):
+            return quad_impl(params[0], params[1], None, *params[2])
+        elif not isinstance(params[1], tuple) and not isinstance(params[2], tuple):
+            return quad_impl(params[0], params[1], None, params[2], None)
+
+    assert False
+
+def _test(params):
+    with test_util.session_context(Data.graph) as sess:
+        _execute_func_on_params(params[1:], 'compile')
+
+        analytic = expectation(*params)
+        quad = _quadrature_expectation(params)
+        analytic, quad = sess.run([analytic, quad])
+        np.testing.assert_almost_equal(quad, analytic, decimal=2)
+
+        _execute_func_on_params(params[1:], 'clear')
+
+
+@pytest.mark.parametrize("distribution", [Data.gauss, Data.gauss_diag])
+@pytest.mark.parametrize("kern", [Data.lin_kern, Data.rbf, Data.rbf_lin_sum, Data.rbf_prod_seperate_dims])
+@pytest.mark.parametrize("feat", [Data.ip])
+@pytest.mark.parametrize("arg_filter", [
+                            lambda p, k, f: (p, k),
+                            lambda p, k, f: (p, (k, f)),
+                            lambda p, k, f: (p, (k, f), (k, f))])
+def test_psi_stats(distribution, kern, feat, arg_filter):
+    params = arg_filter(distribution, kern, feat)
+    _test(params)
+
+
+@pytest.mark.parametrize("distribution", [Data.gauss])
+@pytest.mark.parametrize("mean1", [Data.lin, Data.iden, Data.const, Data.zero])
+@pytest.mark.parametrize("mean2", [Data.lin, Data.iden, Data.const, Data.zero])
+@pytest.mark.parametrize("arg_filter", [
+                            lambda p, m1, m2: (p, m1),
+                            lambda p, m1, m2: (p, m1, m2)])
+def test_mean_function_expectations(distribution, mean1, mean2, arg_filter):
+    params = arg_filter(distribution, mean1, mean2)
+    _test(params)
+
+
+@pytest.mark.parametrize("distribution", [Data.gauss])
+@pytest.mark.parametrize("mean", [Data.lin, Data.iden, Data.const, Data.zero])
+@pytest.mark.parametrize("kern", [Data.rbf, Data.lin_kern])
+@pytest.mark.parametrize("feat", [Data.ip])
+@pytest.mark.parametrize("arg_filter", [
+                            lambda p, k, f, m: (p, (k, f), m),
+                            lambda p, k, f, m: (p, m, (k, f))])
+def test_kernel_mean_function_expectation(distribution, mean, kern, feat, arg_filter):
+    params = arg_filter(distribution, kern, feat, mean)
+    _test(params)
+
+
+@pytest.fixture(params=[Data.rbf, Data.lin_kern])
+def compile_params(request):
+    with test_util.session_context(Data.graph) as sess:
```

Comment:
`test_util.session_context` works as decorator as well. Therefore, you can write something like this:
```python
@session_context()
def test_eKdiag_no_uncertainty(compile_params):
    sess = tf.get_default_session()
    ...
```